File: 2015/day10.cpp
// https://adventofcode.com/2015/day/10

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>

void look_and_say(Vec<u32> const& input, Vec<u32>& output) {
  output.clear();
  auto index = 0u;
  while (index < input.size()) {
    let current = input[index];
    let count = static_cast<u32>(
        stdr::distance(input |
                       stdv::drop(index) |
                       stdv::take_while(aoc::equal_to_value{current})));
    output.push_back(count);
    output.push_back(current);
    index += count;
  }
}

template <int iterations>
int solve_case(const std::string& filename) {
  auto input = aoc::read_single_line(filename) |
               aoc::views::to_number<u32>() |
               aoc::collect_vec<u32>();
  auto output = decltype(input){};
  for ([[maybe_unused]] auto index : stdv::iota(0, iterations)) {
    look_and_say(input, output);
    std::swap(input, output);
  }
  return input.size();
}

int main() {
  std::println("Part 1");
  AOC_EXPECT_RESULT(237746, solve_case<40>("day10.example"));
  AOC_EXPECT_RESULT(360154, solve_case<40>("day10.input"));
  std::println("Part 2");
  AOC_EXPECT_RESULT(3369156, solve_case<50>("day10.example"));
  AOC_EXPECT_RESULT(5103798, solve_case<50>("day10.input"));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2015/src/day10.rs
fn look_and_say(input: &Vec<u32>, output: &mut Vec<u32>) {
    output.clear();
    let mut index = 0;
    while index < input.len() {
        let current = input[index];
        let mut count = 0usize;
        input
            .iter()
            .skip(index)
            .take_while(|&&x| x == current)
            .for_each(|_| count += 1);
        output.push(count as u32);
        output.push(current);
        index += count;
    }
}

fn parse(filename: &str) -> Vec<u32> {
    std::fs::read_to_string(filename)
        .unwrap()
        .trim()
        .chars()
        .map(|c| c.to_digit(10).unwrap())
        .collect()
}

fn solve_case<const ITERATIONS: usize>(digits: &[u32]) -> usize {
    let mut input = digits.to_vec();
    let mut output = Vec::new();
    for _ in 0..ITERATIONS {
        look_and_say(&input, &mut output);
        std::mem::swap(&mut input, &mut output);
    }
    input.len()
}

fn main() {
    println!("Part 1");
    let example = parse("day10.example");
    aoc::expect_result!(237746, solve_case::<40>(&example));
    let input = parse("day10.input");
    aoc::expect_result!(360154, solve_case::<40>(&input));

    println!("Part 2");
    aoc::expect_result!(3369156, solve_case::<50>(&example));
    aoc::expect_result!(5103798, solve_case::<50>(&input));
}
